### PR TITLE
Add intelligent Sequence StorageNode creation using registered NodeSe…

### PR DESCRIPTION
**Pull request by Sunderlandkyl**
…quencers

### CODE REVIEW

**Comment by Sunderlandkyl** [comment]
@lassoan Any changes to make?

File: Sequences/MRML/vtkMRMLSequenceNode.cxx [hunk 1]
```diff
@@ -646,24 +646,22 @@ std::string vtkMRMLSequenceNode::GetDefaultStorageNodeClassName(const char* file
...
+  // Use specific sequence storage node, if possible
+  vtkMRMLNodeSequencer::NodeSequencer* sequencer = vtkMRMLNodeSequencer::GetInstance()->GetNodeSequencer(this->GetNthDataNode(0));
+  std::string storageNodeClassName = sequencer->GetDefaultSequenceStorageNodeClassName();
+  vtkSmartPointer<vtkMRMLStorageNode> storageNode = vtkSmartPointer<vtkMRMLStorageNode>::Take(
+    vtkMRMLStorageNode::SafeDownCast(this->GetScene()->CreateNodeByClass(storageNodeClassName.c_str())));
+
+  // Filename is not specified or it is specified and there is a supported file extension
+  if (!filename || (filename && !storageNode->GetSupportedFileExtension(filename, false, true).empty()))
+  {
+    // cannot write into the required file extension
```
Comment (lassoan) [inline Sequences/MRML/vtkMRMLSequenceNode.cxx]: remove this comment (it was for something else)

File: Sequences/MRML/vtkMRMLNodeSequencer.cxx [hunk 2]
```diff
@@ -54,6 +54,7 @@ vtkMRMLNodeSequencer::NodeSequencer::NodeSequencer()
   this->RecordingEvents = vtkSmartPointer< vtkIntArray >::New();
   this->RecordingEvents->InsertNextValue(vtkCommand::ModifiedEvent);
   this->SupportedNodeClassName = "vtkMRMLNode";
+  this->SequenceStorageNodeClassName = "vtkMRMLSequenceStorageNode";
```
Comment (lassoan) [inline Sequences/MRML/vtkMRMLNodeSequencer.cxx]: Should not this be this->**Default**SequenceStorageNodeClassName?

File: Sequences/MRML/vtkMRMLSequenceNode.cxx [hunk 3]
```diff
@@ -646,24 +646,22 @@ std::string vtkMRMLSequenceNode::GetDefaultStorageNodeClassName(const char* file
...
-  for (std::vector< vtkSmartPointer<vtkMRMLStorageNode> >::iterator specializedStorageNodeIt = specializedStorageNodes.begin();
-    specializedStorageNodeIt != specializedStorageNodes.end(); ++specializedStorageNodeIt)
-  {
-    if (filename && (*specializedStorageNodeIt)->GetSupportedFileExtension(filename, false, true).empty())
-    {
-      // cannot write into the required file extension
-      continue;
-    }
-    if ((*specializedStorageNodeIt)->CanWriteFromReferenceNode(this))
+  // Use specific sequence storage node, if possible
```
Comment (lassoan) [inline Sequences/MRML/vtkMRMLSequenceNode.cxx:657]: It is strange that vtkMRMLSequenceNode::CreateDefaultStorageNode() does not use GetDefaultStorageNodeClassName. Could we completely remove vtkMRMLSequenceNode::CreateDefaultStorageNode() method?

**Review by lassoan — COMMENTED** [review]
There are a few minor questions but overall looks good. See details in in-line comments.

File: Sequences/MRML/vtkMRMLSequenceNode.cxx (hunk 3 above)
Comment (Sunderlandkyl) [reply Sequences/MRML/vtkMRMLSequenceNode.cxx:657]: Yes, I think that it could be removed.

Comment (Sunderlandkyl) [reply Sequences/MRML/vtkMRMLSequenceNode.cxx:657]: Apparently the method is still needed. Default implementation returns NULL:
https://github.com/Slicer/Slicer/blob/4d4b63d4a0a00f8af0ca9ffa82140e17387729a2/Libs/MRML/Core/vtkMRMLStorableNode.cxx#L374-L378

Comment (Sunderlandkyl) [reply Sequences/MRML/vtkMRMLSequenceNode.cxx:657]: @lassoan Would this be because CreateDefaultStorageNode() is supposed to work even without a scene?

**Comment by Sunderlandkyl** [comment]
@lassoan I've added a test as well to check that the correct storage nodes are being used.

**Pull request merged**